File: DataStructure/Stack/Stack.c
/*
Note:

Keep the usage of in-built function to the minimum
addOneEle/Optimize  features
Please follow best coding practices
addOneEle comments if necessary
And make the Code Bug free

*/

// C program for array implementation of stack 
#include <limits.h> 
#include <stdio.h> 
#include <stdlib.h> 

// A structure to represent a stack 
struct Stack { 
	int top; 
	unsigned capacity; 
	int* array; 
}; 

// function to create a stack of given capacity. It initializes size of 
// stack as 0 
struct Stack* createStack(unsigned capacity) 
{ 
	struct Stack* stack = (struct Stack*)malloc(sizeof(struct Stack)); 
	stack->capacity = capacity; 
	stack->top = -1; 
	stack->array = (int*)malloc(stack->capacity * sizeof(int)); 
	return stack; 
} 

// Stack is full when top is equal to the last index 
int isFull(struct Stack* stack) 
{ 
	return stack->top == stack->capacity - 1; 
} 

// Stack is empty when top is equal to -1 
int isEmpty(struct Stack* stack) 
{ 
	return stack->top == -1; 
} 

// Function to add an item to stack. It increases top by 1 
void push(struct Stack* stack, int item) 
{ 
	if (isFull(stack)) 
		return; 
	stack->array[++stack->top] = item; 
	printf("Pushed: %d \n", item); 
} 

// Function to remove an item from stack. It decreases top by 1 
int pop(struct Stack* stack) 
{ 
	if (isEmpty(stack)) 
		return INT_MIN; 
	return stack->array[stack->top--]; 
} 

// Function to return the top from stack without removing it 
int peek(struct Stack* stack) 
{ 
	if (isEmpty(stack)) 
		return INT_MIN; 
	return stack->array[stack->top]; 
} 

//Function to print stack elements
void print_stack(struct Stack* stack){
	int i=0;
	int top = stack->top;
	printf("Stack: ");
    for(i=top;i>=0;i--){
		printf("%d ", stack->array[i]);
	}
    
    printf("\n");
    }

// Driver program to test above functions 
int main() 
{ 
	struct Stack* stack = createStack(100); 

	print_stack(stack);
	push(stack, 10); 
    print_stack(stack);
	push(stack, 20); 
    print_stack(stack);
	printf("Popped: %d\n", pop(stack)); 
	print_stack(stack);

	return 0; 
}

File: DataStructure/Stack/Stack.py
# Note:

# Keep the usage of in-built function to the minimum
# addOneEle/Optimize  features
# Please follow best coding practices
# addOneEle comments if necessary
# And make the Code Bug free

#create a class named stack
class Stack:
    def __init__(self):
        self.items = []

	# to check weather the stack is empty
    def isEmpty(self):
        return self.items == []

	# to push the element to the stack
    def push(self, item):
        print("Pushed: ",item)
        self.items.append(item)

	# to pop the element out of the stack
    def pop(self):
        print("Popped: ",end="")
        return self.items.pop()

	# to return top element of the stack 
    def peek(self):
        return self.items[len(self.items)-1]

	#to return the size of the stack
    def size(self):
        return len(self.items)

	#to print the stack elements
    def print_stack(self):
        print("Stack: ",end="")
        for i in range(0, len(self.items)):
            print(self.items[len(self.items)-i-1], end=" ")
        print()

#driver code
s=Stack()
s.print_stack()
print("The size of the stack is", s.size())
s.push(10)
s.print_stack()
print("The size of the stack is", s.size())
s.push(20)
s.print_stack()
print("The size of the stack is", s.size())
print("The top element is", s.peek())
print(s.pop())
s.print_stack()
print("The size of the stack is", s.size())
